eth/downloader/downloader.go

```diff
@@ -897,31 +1005,42 @@ func (d *Downloader) fetchBlocks(from uint64) error {
 		case <-d.cancelCh:
 			return errCancelBlockFetch
 
-		case blockPack := <-d.blockCh:
+		case <-d.hashCh:
+			// Out of bounds eth/61 hashes received, ignore them
+
+		case <-d.blockCh:
+			// Out of bounds eth/61 blocks received, ignore them
+
+		case bodyPack := <-d.bodyCh:
 			// If the peer was previously banned and failed to deliver it's pack
 			// in a reasonable time frame, ignore it's message.
-			if peer := d.peers.Peer(blockPack.peerId); peer != nil {
-				// Deliver the received chunk of blocks, and demote in case of errors
-				err := d.queue.Deliver(blockPack.peerId, blockPack.blocks)
+			if peer := d.peers.Peer(bodyPack.peerId); peer != nil {
+				// Deliver the received chunk of bodies, and demote in case of errors
+				err := d.queue.Deliver(bodyPack.peerId, bodyPack.transactions, bodyPack.uncles)
 				switch err {
 				case nil:
 					// If no blocks were delivered, demote the peer (need the delivery above)
-					if len(blockPack.blocks) == 0 {
+					if len(bodyPack.transactions) == 0 || len(bodyPack.uncles) == 0 {
 						peer.Demote()
 						peer.SetIdle()
-						glog.V(logger.Detail).Infof("%s: no blocks delivered", peer)
+						glog.V(logger.Detail).Infof("%s: no block bodies delivered", peer)
 						break
 					}
 					// All was successful, promote the peer and potentially start processing
 					peer.Promote()
 					peer.SetIdle()
-					glog.V(logger.Detail).Infof("%s: delivered %d blocks", peer, len(blockPack.blocks))
+					glog.V(logger.Detail).Infof("%s: delivered %d:%d block bodies", peer, len(bodyPack.transactions), len(bodyPack.uncles))
 					go d.process()
 
 				case errInvalidChain:
 					// The hash chain is invalid (blocks are not ordered properly), abort
 					return err
 
+				case errInvalidBody:
+					// The peer delivered something very bad, drop immediately
+					glog.V(logger.Error).Infof("%s: delivered invalid block, DROP DROP DROP (i.e. implement)!!!", peer)
```

Comment:
Ah, yeah. I left it out as I thought it requires a bit more work. Turns out the downloader already has the needed logic for it since I refactored it a while ago so it's just a single line of code :D. Will write an additional test to make sure it works as intended and push the commit.